src/securityPolicy.c

```diff
@@ -634,51 +636,81 @@ security_policy_t policyForSignTxStakePoolRegistrationConfirm(
 }
 
 // For each withdrawal
-security_policy_t policyForSignTxWithdrawal()
+security_policy_t policyForSignTxWithdrawal(
+        sign_tx_usecase_t signTxUsecase,
+        const stake_credential_type_t stakeCredentialType,
+		const bip44_path_t* stakingKeyPath
+)
 {
-	// No need to check withdrawals
-	SHOW();
+	switch (signTxUsecase) {
+	case SIGN_TX_USECASE_ORDINARY_TX:
+		DENY_UNLESS(stakeCredentialType == STAKE_CREDENTIAL_KEY_PATH);
+		ASSERT(stakingKeyPath != NULL);
+		DENY_UNLESS(bip44_isOrdinaryStakingKeyPath(stakingKeyPath));
+		ALLOW();
+		break;
+
+	case SIGN_TX_USECASE_MULTISIG:
+		DENY_UNLESS(stakeCredentialType == STAKE_CREDENTIAL_SCRIPT_HASH);
+		ASSERT(stakingKeyPath == NULL);
+		ALLOW();
+		break;
+
+	default:
+		ASSERT(false);
+	}
+
+	DENY(); // should not be reached
 }
 
 
+static inline security_policy_t nonPoolSubPolicy(const bip44_path_t* pathSpec)
+{
+	switch (bip44_classifyPath(pathSpec)) {
+	case PATH_ORDINARY_SPENDING_KEY:
+	case PATH_ORDINARY_STAKING_KEY:
+	case PATH_MULTISIG_SPENDING_KEY:
+	case PATH_MULTISIG_STAKING_KEY:
+	case PATH_POOL_COLD_KEY:
+		if (bip44_isPathReasonable(pathSpec)) {
+			ALLOW();
+		} else {
+			WARN();
+		}
+		break;
+
+	case PATH_MINT_KEY:
+		SHOW();
+
+	default:
+		DENY();
+		break;
+	}
+}
+
 // For each transaction witness
 // Note: witnesses reveal public key of an address
 // and Ledger *does not* check whether they correspond to
 // previously declared inputs and certificates
 security_policy_t policyForSignTxWitness(
```

Comment:
I imagine that this function should only consist of a switch statement that distinguishes the usecase and then for each use case, we call a separate function that would contain a switch statement distinguishing cases for the `bip44_classifyPath` result. Then it would be straightforward and easily auditable.

Reply:
That's basically what is happening right now, but we have TWO things to decide (deny/allow and show/prompt/warn). Right now I separated those two for the non-pool use cases, and I think it looks better for it. But I'm open for ideas.